### DISCUSSION TITLE
After Win+A was pressed, Control Center icon disappears at some point: possible workaround?

### DISCUSSION THREAD

**Discussion by plofhaan** [valinet/ExplorerPatcher#242] [Q&A] [answered]
I managed to work around this problem as follows:
In ExplorerPatcher settings, keep the "Hide Control Center button" setting set to _enabled_.
After Win+A was pressed, left-click on any system tray icon to make the Control Center icon become visible.

The Control Center icon also becomes visible if I cycle (disable/enable) the "Show Desktop button" setting of ExplorerPatcher. Another way to make this icon become visible is to hover the mouse over the bottom part of the Quick Settings menu that appears when Win+A is pressed (the brightness/volume sliders), and then just wait a few seconds. It's as if some kind of GUI interaction is needed to trigger the refresh. But each time when Explorer starts, I have to force the icon to become visible again so I was hoping for a more permanent solution.

**Comment by valinet** [comment]
So... I don't understand what you describe here. What's the actual problem? What workaround is needed? Workaround for what? For me, it works as it should: hiding it hides the icon but `Win`+`A` is still usable, showing it shows the icon and `Win`+`A` is still usable, so...

Indeed, for `Win`+`A` to be available while the icon is hidden, I have to briefly toggle its visibility, show the pop-up, and then hide it again, otherwise Control Center won't show. Anyway, maybe that glitches out from time to time, but I don't know any other solution to achieve that, so maybe keeping it enabled all the time is a workaround if you use the thing so often, idk... I can't do much about that as I don't know what else to try... maybe have a look on the source, attempt a few things and see what results you obtain...?

Anyway, just as an aside, for a stupid tray icon, this generates much more discussion that other things that would seem logical to irritate people more; for example, no one complains about the battery flyout which is broken, which you'd think more people use than than people wanting to do something about this stupid icon. It's true, that for the battery flyout, much better alternatives are recommended in the README, which basically makes fixing it useless anyway.

**Reply by plofhaan** [reply]
Thank you for your reply. :) I don't know why my gear wheel disappears, maybe it's because I have installed TranslucentTB app from the Microsoft Store. I really don't know. I also use Open-Shell and Battery Mode. Anyway, I will try to be more specific of what exactly I am experiencing.

With the "Hide Control Center button" set to _disabled_, the gear wheel is indeed visible, but if I press Win+A, and, next, I hover my mouse over the gear wheel, the gear wheel disappears. To be more specific, it disappears in two steps. The first step is: immediately it jumps to the left in such a way that the left half of the gear wheel is gone and I can no longer click on it, but the tooltip still works when I hover my mouse over the (right half of the) gear wheel, and, Win+A still works. The second step occurs when I left-click on a system tray icon (any system tray icon), which is when all my system tray icons immediately jump one icon position to the right so, as a result from this, the gear wheel disappears (completely, that is). But yes, Win+A still continues to work like normal. The workaround that I have described is how I can prevent the gear wheel from disappearing, it works, at least for me.

With the "Hide Control Center button" set to _enabled_, the gear wheel is hidden, but if I press Win+A, and, next, I click on a system tray icon, the gear wheel un-hides itself. After that, I can choose between my keyboard (Win+A) or my mouse (left-click on the gear wheel) to bring up the Control Center menu, without any further problems. I also found that another successful workaround is to press Win+A two times in a row, but this workaround is successful only if it is performed a few seconds before all my system tray icons have appeared. On my laptop after Win11 logs on it takes well over half a minute to load them all. (Yeah... I know, lol) This method (of workaround) effectively saves the extra effort of having to click on one of my system tray icons after this. If either I do it too soon or I do it too late, it misses that effect, though. I don't have Visual Studio atm so I can't mess with your source code, but I made an .exe file with the VB compiler from my .NET Framework. I have created a task in task scheduler to run this at logon. Of course this is just a primitive experiment. All I had to do to get it working was to keep adjusting the wait duration via trial and error. But for me it appears to work. :D So here is the code:

```
Imports System.Runtime.InteropServices
Imports System.Windows.Forms

Public Module GearWheelKeeper

   Const KEYEVENTF_KEYDOWN As Integer = &H0
   Const KEYEVENTF_KEYUP As Integer = &H2

   Declare Sub keybd_event Lib "User32" (ByVal bVk As Byte, ByVal bScan As Byte, ByVal dwFlags As UInteger, ByVal dwExtraInfo As UInteger)

   Private Sub wait(ByVal hundredths As Integer)
      For i As Integer = 0 To hundredths
         System.Threading.Thread.Sleep(10)
         Application.DoEvents()
      Next
   End Sub
   Private Sub wina
      keybd_event(CByte(Keys.LWin), 0, KEYEVENTF_KEYDOWN, 0) 'press the left Win key down
      keybd_event(CByte(Keys.A), 0, KEYEVENTF_KEYDOWN, 0) 'press the A key down
      keybd_event(CByte(Keys.A), 0, KEYEVENTF_KEYUP, 0) 'release the A key
      keybd_event(CByte(Keys.LWin), 0, KEYEVENTF_KEYUP, 0) 'release the left Win key
   End Sub
   Public Sub Main()
      wait(1700)
      wina
      wait(80)
      wina
   End Sub
End Module
```

**Reply by plofhaan** [reply]
GearWheelKeeper 2.0:
```
Imports System.Runtime.InteropServices
Imports System.Windows.Forms
Imports System.Drawing

Public Module GearWheelKeeper

   Const KEYEVENTF_KEYDOWN As Integer = &H0
   Const KEYEVENTF_KEYUP As Integer = &H2

   Declare Sub keybd_event Lib "User32" (ByVal bVk As Byte, ByVal bScan As Byte, ByVal dwFlags As UInteger, ByVal dwExtraInfo As UInteger)

   Private Sub wait(ByVal hundredths As Integer)
      For i As Integer = 0 To hundredths
         System.Threading.Thread.Sleep(10)
         Application.DoEvents()
      Next
   End Sub
   Private Sub wina
      keybd_event(CByte(Keys.LWin), 0, KEYEVENTF_KEYDOWN, 0) 'press the left Win key down
      keybd_event(CByte(Keys.A), 0, KEYEVENTF_KEYDOWN, 0) 'press the A key down
      keybd_event(CByte(Keys.A), 0, KEYEVENTF_KEYUP, 0) 'release the A key
      keybd_event(CByte(Keys.LWin), 0, KEYEVENTF_KEYUP, 0) 'release the left Win key
   End Sub
   Public Class AppContext
      Inherits ApplicationContext
      Private notifyIcon As NotifyIcon
      Public Sub New()
         AddHandler Application.ApplicationExit, AddressOf OnApplicationExit
         notifyIcon = New NotifyIcon()
         notifyIcon.Icon = New Icon(SystemIcons.Application, 40, 40)
         AddHandler notifyIcon.MouseClick, AddressOf OnIconMouseClick
         notifyIcon.Visible = True
         OnIconMouseClick(notifyIcon, New MouseEventArgs(MouseButtons.Left, 1, 0, 0, 0))
      End Sub
      Private Sub OnApplicationExit(ByVal sender As Object, ByVal e As EventArgs)
         If notifyIcon IsNot Nothing Then
            notifyIcon.Dispose()
         End If
      End Sub
      Private Sub OnIconMouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
         notifyIcon.Visible = False
         notifyIcon.Icon = New Icon("%windir%\System32\shell32.dll,-50")
         ExitThread()
      End Sub
   End Class

   Public Sub Main()
      wait(500)
      wina
      wait(80)
      wina
      wait(80)
      Dim ctx As New AppContext()
      Application.Run(ctx)
   End Sub
End Module
```

**Comment by plofhaan** [comment, marked as answer]
This issue appears to be fixed in latest ExplorerPatcher version. Thanks!

**Reply by valinet** [reply]
That’s great to hear, yeah, I added some code to fix it up. Thanks.
